### ISSUE TITLE
Endpoint build can take too long for CNI

### ISSUE THREAD

**Issue by tgraf** [cilium/cilium#5028]
```
Events:
  Type     Reason                  Age              From               Message
  ----     ------                  ----             ----               -------
  Normal   SuccessfulMountVolume   3m               kubelet, minikube  MountVolume.SetUp succeeded for volume "cilium-unix-sock-dir"
  Normal   SuccessfulMountVolume   3m               kubelet, minikube  MountVolume.SetUp succeeded for volume "istio-envoy"
  Normal   SuccessfulMountVolume   3m               kubelet, minikube  MountVolume.SetUp succeeded for volume "default-token-vptwr"
  Normal   SuccessfulMountVolume   3m               kubelet, minikube  MountVolume.SetUp succeeded for volume "istio-certs"
  Normal   Scheduled               3m               default-scheduler  Successfully assigned details-v1-5cdff44756-98mhs to minikube
  Warning  FailedCreatePodSandBox  3m               kubelet, minikube  Failed create pod sandbox: rpc error: code = Unknown desc = NetworkPlugin cni failed to set up pod "details-v1-5cdff44756-98mhs_default" network: Unable to create endpoint: [PUT /endpoint/{id}][500] putEndpointIdFailed  endpoint 21062 did not synchronously regenerate after timeout
  Warning  FailedCreatePodSandBox  2m               kubelet, minikube  Failed create pod sandbox: rpc error: code = Unknown desc = NetworkPlugin cni failed to set up pod "details-v1-5cdff44756-98mhs_default" network: Unable to create endpoint: [PUT /endpoint/{id}][500] putEndpointIdFailed  endpoint 51849 did not synchronously regenerate after timeout
  Warning  FailedCreatePodSandBox  2m (x2 over 2m)  kubelet, minikube  Failed create pod sandbox: rpc error: code = Unknown desc = NetworkPlugin cni failed to set up pod "details-v1-5cdff44756-98mhs_default" network: Unable to create endpoint: context deadline exceeded
  Warning  FailedCreatePodSandBox  1m               kubelet, minikube  Failed create pod sandbox: rpc error: code = Unknown desc = NetworkPlugin cni failed to set up pod "details-v1-5cdff44756-98mhs_default" network: Unable to create endpoint: [PUT /endpoint/{id}][500] putEndpointIdFailed  endpoint 33978 did not synchronously regenerate after timeout
  Normal   SandboxChanged          1m (x5 over 3m)  kubelet, minikube  Pod sandbox changed, it will be killed and re-created.
  Normal   Started                 1m               kubelet, minikube  Started container
  Normal   Created                 1m               kubelet, minikube  Created container
  Normal   Pulled                  1m               kubelet, minikube  Container image "busybox" already present on machine
  Normal   Pulled                  57s              kubelet, minikube  Container image "docker.io/istio/proxy_init:0.8.0" already present on machine
  Normal   Created                 57s              kubelet, minikube  Created container
  Normal   Started                 56s              kubelet, minikube  Started container
  Normal   Pulling                 55s              kubelet, minikube  pulling image "docker.io/istio/examples-bookinfo-details-v1:0.2.8"
  Normal   Pulled                  24s              kubelet, minikube  Successfully pulled image "docker.io/istio/examples-bookinfo-details-v1:0.2.8"
  Normal   Created                 24s              kubelet, minikube  Created container
  Normal   Started                 24s              kubelet, minikube  Started container
  Normal   Pulling                 24s              kubelet, minikube  pulling image "docker.io/cilium/istio_proxy_debug:0.8.0"
```

**Comment by tgraf** [comment]
Related #4995

**Comment by ianvernon** [comment]
We can bump the timeout on the client for CNI:
https://github.com/cilium/cilium/blob/7a10b865c278c04a942a1ae31220825639c82d66/pkg/client/client.go#L79-L106

Is it possible for us to have some default timeout, then increase it if the first request fails / implement retry logic in CNI add? The issue with synchronous add is that we are entirely at the behest of how much load is on the system; if Cilium is resource constrained, builds will almost certainly fail.

**Comment by tgraf** [comment]
> Is it possible for us to have some default timeout, then increase it if the first request fails / implement retry logic in CNI add?

That's an excellent idea. We still need a hard cap though.

**Comment by tgraf** [comment]
Resolved by #4998 

Regarding the suggestion from @ianvernon. It is a great idea but difficult to implement as the client side timeout has to be increased as well. Let's see if this is still ever an issue after the timeout bump and then discuss again.
